rust/relay/ebpf-turn-router/src/move_payload.rs

```diff
@@ -0,0 +1,245 @@
+use aya_ebpf::{
+    cty::c_void,
+    helpers::{bpf_xdp_adjust_tail, bpf_xdp_load_bytes, bpf_xdp_store_bytes},
+    programs::XdpContext,
+};
+use network_types::{
+    eth::EthHdr,
+    ip::{Ipv4Hdr, Ipv6Hdr},
+    udp::UdpHdr,
+};
+
+use crate::{channel_data::CdHdr, error::Error};
+
+/// Public wrappers
+#[inline(always)]
+pub fn add_channel_data_header_ipv4(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    move_payload::<{ Ipv4Hdr::LEN }, { -(CdHdr::LEN as i32) }>(ctx, Some(&header))
+}
+
+#[inline(always)]
+pub fn add_channel_data_header_ipv6(ctx: &XdpContext, header: CdHdr) -> Result<(), Error> {
+    move_payload::<{ Ipv6Hdr::LEN }, { -(CdHdr::LEN as i32) }>(ctx, Some(&header))
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv4(ctx: &XdpContext) -> Result<(), Error> {
+    move_payload::<{ Ipv4Hdr::LEN }, { CdHdr::LEN as i32 }>(ctx, None)
+}
+
+#[inline(always)]
+pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
+    move_payload::<{ Ipv6Hdr::LEN }, { CdHdr::LEN as i32 }>(ctx, None)
+}
+
+/// Single entry point to add/remove bytes at the start of the UDP payload using only adjust_tail.
+/// - `delta < 0`: add `-delta` bytes (extend tail, slide payload toward tail, optionally write header)
+/// - `delta > 0`: remove `delta` bytes (slide payload toward head, then shrink tail)
+#[inline(always)]
+pub fn move_payload<const IP_HEADER_LEN: usize, const DELTA: i32>(
+    ctx: &XdpContext,
+    header: Option<&CdHdr>,
+) -> Result<(), Error> {
+    if DELTA < 0 {
+        slide_backward_and_extend::<IP_HEADER_LEN, DELTA>(ctx, header)
+    } else {
+        slide_forward_and_shrink::<IP_HEADER_LEN, DELTA>(ctx)
+    }
+}
+
+/* -------------------- Implementation Helpers -------------------- */
+
+#[inline(always)]
+fn payload_offset<const IP_HEADER_LEN: usize>() -> usize {
+    EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN
+}
+
+/// Small, stack-safe copy: N must be <= 64 (kept out-of-line to avoid stack blow-up).
+#[inline(never)]
+fn copy_n<const N: usize>(ctx: &XdpContext, src: usize, dst: usize) -> Result<(), Error> {
```

Comment:
Please define functions in the order they are used, i.e. helpers at the very bottom.

I'd like to be able to read from top to bottom of the file with an increase in level of detail for what is happening and not come across functions that haven't been used yet (unless they are public)